pacemaker-remoted load_env_var refactors, and assorted minor changes

nrwahl2 opened the pull request:
I'm trying to fix the Coverity issues that are popping up for #3832 and these are also appearing on my machine. There are some more I haven't reached yet.

clumens commented on daemons/execd/remoted_pidone.c:
```diff
@@ -94,102 +94,116 @@ find_env_var_name(char *line, char **first, char **last)
-load_env_vars(const char *filename)
+load_env_vars(void)
 {
+    static const char *const filename = "/etc/pacemaker/pcmk-init.env";
```
If the filename is always going to be the same, could we just make it a #define and not have to use this ridiculous type?

nrwahl2 replied:
Oh yeah, this is a combination of (a) remnant of another approach and (b) late night delirium

nrwahl2 commented:
Used a define per review (https://github.com/ClusterLabs/pacemaker/compare/17bb059c39e3ee0b75a8fc26c259a26c5241c719..3ebf24e3100e21a2900009ebcb5c5c7edafa92a1)

clumens commented:
Were you able to test the changes to load_env_vars?  That, or some unit tests, would make me feel a whole lot better about all that small batch string processing code.

nrwahl2 commented:
> Were you able to test the changes to load_env_vars? That, or some unit tests, would make me feel a whole lot better about all that small batch string processing code.

Not really. That's gonna take me a little while to set up and figure out (won't be ready today during business hours). But it is a good idea.

Also interested in whether there's a better way than rolling our own parser, but haven't found anything yet.

nrwahl2 commented:
> Not really. That's gonna take me a little while to set up and figure out (won't be ready today during business hours). But it is a good idea.
> 
> Also interested in whether there's a better way than rolling our own parser, but haven't found anything yet.

I'm planning to use `pcmk__scan_nvpair()` here (now that I'm aware it exists) and then process the key and value separately. It will hopefully make things easier to reason about, even though there will still be corner cases involving escapes and nested quotes for now. See also #3834.

clumens commented:
> I'm planning to use `pcmk__scan_nvpair()` here (now that I'm aware it exists) and then process the key and value separately. It will hopefully make things easier to reason about, even though there will still be corner cases involving escapes and nested quotes for now. See also #3834.

Okay, I'll wait to do final review on this until after you've moved to using `pcmk__scan_nvpair()`.

> I think the authors made the `load_env_vars()` function way too complicated in our habitual obsession with "try to never allocate heap memory for a string." Which makes sense in the kernel, but IMO it's pretty silly in MOST of Pacemaker, where we're allocating large data structures frequently. Way too complicated without even achieving totally correct behavior, because it's so easy to get bogged down in the character-by-character parsing details.

I absolutely agree.  If I were writing this function from scratch, I would (1) have it take file contents as a parameter instead of a file name, (2) put it in a library somewhere as private API to allow us to unit test it, and (3) find something in glib that could be used to do the parsing instead of writing it by hand.  Now we're stuck with having to support whatever bugs there might be in processing since otherwise we'll break people's remote nodes.

nrwahl2 commented:
> Now we're stuck with having to support whatever bugs there might be in processing since otherwise we'll break people's remote nodes.

Not necessarily. Obviously we've run into many situations where users are relying on behavior they shouldn't be... but these corner cases seem like pretty clear bugs that shouldn't be guaranteed to work, and that it's unlikely anyone is using. I'd be willing to gamble, if we actually feel like fixing them.

That's not something I want to spend time on right now though. If we find a good parser in a library, fine. But setting flags for quote nesting levels, whether we just saw a backslash and are escaping the next character, etc., sounds really unappealing.

nrwahl2 commented:
Rebased on current main, which led to dropping the final commit. Otherwise no changes.

clumens commented on daemons/execd/remoted_schemas.c:
```diff
@@ -45,41 +49,27 @@ static void
 
-        } else {
+    // rc == -1: nftw() itself failed
+    if (errno == ENOENT) {
```
Can we actually rely on errno here?  The nftw man page doesn't mention setting it, nor does it mention any `E*` error codes.  I think what we're counting on is that errno from the remove() call in `rm_files` will still be valid, and who knows what could happen to it between here and there.

nrwahl2 replied:
I think so, but that's a very good question.
* `nftw()` is a C standard library function, and according to the C standard it's not allowed to set `errno` to 0.
* On my laptop, `nftw()` is documented as making a `stat()` system call. It would be reasonable to assume that we keep the `errno` from `stat()`, though this is not stated explicitly as I'd prefer it to be.
* The POSIX standard says: "[If t]he nftw() function detects an error other than [EACCES]... nftw() shall return -1 and set errno to indicate the error." (https://pubs.opengroup.org/onlinepubs/9699919799/)

My testing used the following file:
```
static int
rm_files(const char *fpath, const struct stat *sb, int typeflag,
         struct FTW *ftwbuf)
{
    return 0;
}

int main(void)
{
    int rc = nftw("/asdf", rm_files, 10,
                  FTW_DEPTH|FTW_MOUNT|FTW_PHYS);

    if (rc == -1) {
        printf("error: %s\n", strerror(errno));
    }
    return 0;
}
```

The POSIX doc is the most compelling piece of this really, but Linux isn't guaranteed to comply with POSIX.

If we wanted to be especially paranoid, we could have a file-scope boolean variable that we set to false before we call `nftw()` and set to true within `rm_files()`, to indicate that we ever called `rm_files()`.

nrwahl2 replied:
> I think what we're counting on is that errno from the remove() call in rm_files will still be valid

No, as shown in my code snippet above (which prints "No such file or directory"), the `rm_files()` function is irrelevant when the base directory doesn't exist.

nrwahl2 commented:
I added a commit to fix the `rand()` suppression. I also split up the `load_env_vars()` stuff into several more incremental commits and then used `pcmk__scan_nvpair()` in the final one. It doesn't really make that function any shorter, but IMO it's a lot more readable now.

I have not tested yet.

nrwahl2 commented:
Still need to test the `load_env_vars()` stuff and maybe functionize the value validation part, and run cts-lab for a bit to maybe find any obvious issues with the remote message stuff.

I can't easily test the `localized_remote_header()` involving byte swapping for endianness, because I don't have a big-endian machine. The last time I tried to set up a big-endian VM, if I ever succeeded at all, it was an enormous pain. I can't don't see much benefit from unit testing it either: we can trust that the GLib functions do what they're supposed to do, as long as the surrounding logic is correct.

But this actually resolves all the Coverity issues on my laptop. There is a suppression for `rand()`, and there are two suppressions in `crm_resource.c` that will go away as part of my `crm_resource` overhaul PRs. And that's it.

clumens commented on daemons/execd/remoted_pidone.c:
```diff
@@ -156,11 +157,17 @@ load_env_vars(void)
         }
 
-        if ((*end != '\0') && (*end != '#')) {
+        if (*end != '\0') {
```
This kind of function is the perfect candidate for unit tests, though it's hiding somewhere inaccessible at the moment and I would hate to expose it even as private API.

nrwahl2 replied:
Yeah, we could make it library-private for that purpose. It is really annoying to me that static functions can't be unit-tested by cmocka AFAICT.

clumens replied:
Yeah that's just an artifact of symbols not being visible if they're static.  If you really want to test a static function, you can wrap the `static` bit in `#if !defined(PCMK__UNIT_TESTING)`.  We could even go so far as to define something like `PCMK__STATIC` (man I wish we had namespacing) to do that for us.  libcrmcommon at least is already set up to build a whole separate version of the library for unit testing, so we can do whatever we want.  This would be harder with daemons.

clumens commented on daemons/fenced/fenced_commands.c:
```diff
@@ -742,10 +742,13 @@ schedule_stonith_command(async_command_t *cmd, fenced_device_t *device)
+        if (delay_max > delay_base) {
+            // coverity[dont_call] Doesn't matter that rand() is predictable
+            cmd->start_delay += rand() % (delay_max - delay_base + 1);
+        }
```
It feels more like addressing the issue of the range, instead of simplification.

clumens commented on daemons/execd/remoted_pidone.c:
```diff
@@ -113,73 +113,77 @@ load_env_vars(const char *filename)
 
         // Look for valid name immediately followed by equals sign
-        if (find_env_var_name(line, &name, &end) && (*++end == '=')) {
+        if (!find_env_var_name(line, &name, &end) || (*++end != '=')) {
```
Yeah, the later commits in this series definitely go a long way towards this goal.

nrwahl2 commented on daemons/execd/remoted_pidone.c:
```diff
@@ -156,11 +157,17 @@ load_env_vars(void)
         value_end = end;
 
+        // Strip trailing comment beginning with '#'
+        comment = strchr(end, '#');
```
Technically this is wrong in the case of, say, `name=value# asdf`. It turns out Bash assigns sets `name` to `"value#"`, not `"value"`. So the `asdf` would be treated as a second token, not as part of a comment, and the whole line would be invalid.

Using `g_shell_parse_argv()` in a later commit fixes that among other things.

clumens commented:
@nrwahl2 Is there anything else remaining to be done on this PR?

nrwahl2 commented:
> @nrwahl2 Is there anything else remaining to be done on this PR?

Only testing, hopefully.

There is no unit testing yet, which may be overkill now that we rely on a handful of GLib function calls... but more importantly I haven't tested this by creating a bundle and making sure it loads environment variables appropriately. I need to refresh on how to do that.
